powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/OperationBehavior.java

```diff
@@ -0,0 +1,469 @@
+            if (OperationStatusDo.PENDING.equals(operationEntity.getStatus())) {
+
+                // Update failure count, check the failure count and FAIL operation if needed
+                final Long failureCount = operationEntity.getFailureCount() + 1;
+                final Long maxFailureCount = operationEntity.getMaxFailureCount();
+
+                if (failureCount < maxFailureCount) {
+                    operationEntity.setFailureCount(failureCount);
+
+                    final OperationEntity savedEntity = operationRepository.save(operationEntity);
+                    final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
+
+                    OperationUserActionResponse response = new OperationUserActionResponse();
+                    response.setResult(UserActionResult.APPROVAL_FAILED);
+                    response.setOperation(operationDetailResponse);
+                    return response;
+                } else {
+                    operationEntity.setStatus(OperationStatusDo.FAILED);
+                    operationEntity.setTimestampFinalized(currentTimestamp);
+                    operationEntity.setFailureCount(maxFailureCount); // just in case, set the failure count to max value
+
+                    final OperationEntity savedEntity = operationRepository.save(operationEntity);
+                    final OperationDetailResponse operationDetailResponse = convertFromEntity(savedEntity);
+
+                    OperationUserActionResponse response = new OperationUserActionResponse();
+                    response.setResult(UserActionResult.OPERATION_FAILED);
+                    response.setOperation(operationDetailResponse);
+                    return response;
+                }
+
+            } else {
+                throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_INVALID_STATE);
+            }
+
+        } else {
+            throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_NOT_FOUND);
+        }
+    }
+
+    public OperationDetailResponse cancelOperation(OperationCancelRequest request) throws GenericServiceException {
+        final Date currentTimestamp = new Date();
+
+        final String operationId = request.getOperationId();
+        final Optional<OperationEntity> operationOptional = operationRepository.findOperation(operationId);
+        if (operationOptional.isPresent()) {
+            final OperationEntity operationEntity = expireOperation(operationOptional.get(), currentTimestamp);
+            if (OperationStatusDo.PENDING.equals(operationEntity.getStatus())) {
+                operationEntity.setStatus(OperationStatusDo.CANCELED);
+                final OperationEntity savedEntity = operationRepository.save(operationEntity);
+                return convertFromEntity(savedEntity);
+            } else {
+                throw localizationProvider.buildExceptionForCode(ServiceError.OPERATION_INVALID_STATE);
```

Comment:
From style point of view, I think it is better to check for the negative cases in the beginning of the logic. This way two levels of nesting can be removed and the code is easier to read.

Reply:
Makes sense.